src/measure_theory/set_integral.lean

```diff
@@ -0,0 +1,284 @@
+/-
+Copyright (c) 2020 Zhouhang Zhou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zhouhang Zhou
+-/
+
+import measure_theory.bochner_integration
+import measure_theory.indicator_function
+
+/-!
+# Set integral
+
+Integrate a function over a subset of a measure space.
+
+## Main definition
+
+`measurable_on`, `integrable_on`, `integral_on`
+
+## Tags
+
+indicator, characteristic
+-/
+
+noncomputable theory
+open_locale classical topological_space
+open set lattice filter topological_space ennreal emetric measure_theory
+
+set_option class.instance_max_depth 50
+
+universes u v w
+variables {α : Type u} {β : Type v} {γ : Type w}
+
+section measurable_on
+variables [measurable_space α] [measurable_space β] [has_zero β] {s : set α} {f : α → β}
+
+/-- `measurable_on s f` means `f` is measurable over the set `s`. -/
+@[reducible]
+def measurable_on (s : set α) (f : α → β) : Prop := measurable (indicator s f)
+
+lemma measurable_on_univ (hf : measurable f) : measurable_on univ f :=
+hf.if is_measurable.univ measurable_const
+
+lemma measurable.measurable_on (hs : is_measurable s) (hf : measurable f) : measurable_on s f :=
+hf.if hs measurable_const
+
+lemma is_measurable.inter_preimage {B : set β}
+  (hs : is_measurable s) (hB : is_measurable B) (hf : measurable_on s f):
+  is_measurable (s ∩ f ⁻¹' B) :=
+begin
+  rw [measurable_on] at hf,
+  replace hf : is_measurable ((indicator s f)⁻¹' B) := hf B hB,
+  rw indicator_preimage at hf,
+  replace hf := hf.diff _,
+  rwa union_diff_cancel_right at hf,
+  { assume a, simp {contextual := tt} },
+  exact hs.compl.inter (measurable_const.preimage hB)
+end
+
+lemma measurable_on.subset {t : set α} (hs : is_measurable s) (h : s ⊆ t) (hf : measurable_on t f) :
+  measurable_on s f :=
+begin
+  have : measurable_on s (indicator t f) := measurable.measurable_on hs hf,
+  simp only [measurable_on, indicator_indicator] at this,
+  rwa [inter_eq_self_of_subset_left h] at this,
+end
+
+lemma measurable_on.union {β} [measurable_space β] [add_monoid β] {t : set α} {f : α → β}
+  (hs : is_measurable s) (ht : is_measurable t) (hsm : measurable_on s f) (htm : measurable_on t f) :
+  measurable_on (s ∪ t) f :=
+begin
+  assume B hB,
+  show is_measurable ((indicator (s ∪ t) f)⁻¹' B),
+  rw indicator_preimage,
+  refine is_measurable.union _ ((hs.union ht).compl.inter (measurable_const.preimage hB)),
+  simp only [union_inter_distrib_right],
+  exact (hs.inter_preimage hB hsm).union (ht.inter_preimage hB htm)
+end
+
+end measurable_on
+
+section integrable_on
+variables [measure_space α] [normed_group β] {s t : set α} {f g : α → β}
+
+/-- `integrable_on s f` means `f` is integrable over the set `s`. -/
+@[reducible]
+def integrable_on (s : set α) (f : α → β) : Prop := integrable (indicator s f)
+
+lemma integrable_on_iff (h : ∀x, x ∈ s → f x = g x) : integrable_on s f ↔ integrable_on s g :=
```

Comment:
I'd expect name `integrable_on_congr` here, and `integrable_on_iff` or `integrable_on_def` could be `integrable on s f <-> integrable (indicator s f)`.